src/crypto/tls/ech.go

```diff
@@ -0,0 +1,874 @@
+		ech.handle.enc = enc
+		ech.payload = ctx.encrypt(rawCompressedHello)
+		hello.encryptedClientHelloOffered = true
+		hello.encryptedClientHello = ech.marshal()
+
+		// Update the HRR pre-shared-key. This is used to encrypt the second
+		// ClientHelloInner in case the server sends an HRR.
+		c.ech.hrrPsk = ctx.hrrPsk()
+
+		// Record the "random" sent in the ClientHelloInner. This value will be
+		// used for the second ClientHelloInner in case the server sends an HRR.
+		c.ech.hrrInnerRandom = helloInner.random
+
+		// Offer ECH.
+		c.ech.st.Offered = true
+		helloInner.raw = nil
+		hello.raw = nil
+		return hello, helloInner, nil
+	}
+
+	hello = new(clientHelloMsg)
+	*hello = *helloBase
+
+	hpkeSuite, err := echAssembleHpkeCipherSuite(dummyKemId, dummyKdfId, dummyAeadId)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	pk, err := hpkeSuite.KEM.Deserialize(echDummyPublicKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	var dummyEch echClient
+	dummyEch.handle.suite.kdfId = dummyKdfId
+	dummyEch.handle.suite.aeadId = dummyAeadId
+
+	dummyEch.handle.configId = make([]byte, dummyKdfOutputLen)
+	if _, err = io.ReadFull(config.rand(), dummyEch.handle.configId); err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	dummyEch.handle.enc, _, err = hpke.SetupBaseS(hpkeSuite, config.rand(), pk, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	// TODO(cjpatton): Ensure the length of the dummy ciphertext "sticks out".
+	dummyEch.payload = make([]byte, len(hello.marshal())+dummyAeadOverheadLen)
+	if _, err = io.ReadFull(config.rand(), dummyEch.payload); err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	hello.encryptedClientHelloOffered = true
+	hello.encryptedClientHello = dummyEch.marshal()
+
+	// Bypass ECH and provide covertext.
+	c.ech.st.Grease = true
+	hello.raw = nil
+	return hello, nil, nil
+}
+
+func (c *Conn) echAcceptOrBypass(hello *clientHelloMsg) (*clientHelloMsg, error) {
+	config := c.config
+	echProvider := config.ServerECHProvider
+
+	// Determine if ECH was offered.
+	offered := false
+	if config.echCanAccept() && hello.encryptedClientHelloOffered {
+		publicNames, err := echProvider.GetPublicNames()
+		if err != nil {
+			c.sendAlert(alertInternalError)
+			return nil, fmt.Errorf("ech: could not resolve public names: %s", err)
+		}
+		for _, name := range publicNames {
+			if hello.serverName == name {
+				offered = true
+			}
+		}
+	}
+
+	// Decide whether to bypass ECH.
+	if !offered || !config.echCanAccept() ||
+		!hello.encryptedClientHelloOffered {
+		if c.hrrTriggered && c.ech.st.Offered {
+			// Detect illegal second ClientHello.
+			c.sendAlert(alertIllegalParameter)
+			return nil, fmt.Errorf("ech: hrr: bypass after offer")
+		}
+
+		if config.echCanAccept() && hello.encryptedClientHelloOffered {
```

Comment:
Has the syntax of the extension been validated at this point? Should it be? It seems like doing this for GREASE should be required (I don't remember if the draft requires that off hand).

Reply:
(We parse it below, but don't hit that line since we return early on 269)